### PR TITLE
Improved FormatMap: Improves performance by about 4x, or nearly 2x in the worst case

### PR DESCRIPTION
**Pull request by aimuz**
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

Improves performance by about 4x, or nearly 2x in the worst case

```release-note
Improved FormatMap: Improves performance by about 4x, or nearly 2x in the worst case
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
Improves performance by about 4x, or nearly 2x in the worst case

old FormatMap
BenchmarkFormatMap-12             873046                1238 ns/op             384 B/op         13 allocs/op
new FormatMap
BenchmarkFormatMap-12            3665762               327.0 ns/op             152 B/op          3 allocs/op
```

### CODE REVIEW

File: pkg/fieldpath/fieldpath.go [hunk 1]
```diff
@@ -18,26 +18,37 @@ package fieldpath
...
+	sort.Strings(keys)
+	var buf strings.Builder
+	buf.Grow(grow)
+	for _, key := range keys {
+		if buf.Len() > 0 {
+			buf.WriteString("\n")
+		}
+		buf.WriteString(key)
+		buf.WriteByte('=')
+		buf.Write(strconv.AppendQuote(nil, m[key]))
```
Comment (smarterclayton) [inline pkg/fieldpath/fieldpath.go:49]: It’s a little weird here not to take advantage of AppendQuote and avoid another big chunk of allocations.  I’d recommend trying to leverage either the builder or reuse a single buffer.

Given the change, do we have sufficient unit tests to verify no regression?

Comment (aimuz) [reply pkg/fieldpath/fieldpath.go]: Yes, I reproduced the adjustment and added the test

**Comment by aimuz** [comment]
@smarterclayton Hello, do you have time to help me review

File: pkg/fieldpath/fieldpath.go [hunk 2]
```diff
@@ -18,26 +18,40 @@ package fieldpath
...
+		}
+		dst = append(dst, key...)
+		dst = append(dst, '=')
+		dst = strconv.AppendQuote(dst, m[key])
 	}
-	fmtStr = strings.TrimSuffix(fmtStr, "\n")
-
-	return
+	// form strings.Builder{}.String()
+	return *(*string)(unsafe.Pointer(&dst))
```
Comment (liggitt) [inline pkg/fieldpath/fieldpath.go:54]: Thanks for the PR, but I don't think we should switch this to use `unsafe`. From what I can see, this function is not used in a hotpath that justifies the added complexity / safety issues.

Comment (aimuz) [reply pkg/fieldpath/fieldpath.go]: Yes indeed, it does not take into account this situation

File: pkg/fieldpath/fieldpath.go [hunk 3]
```diff
@@ -18,26 +18,38 @@ package fieldpath
...
-	keys := sets.NewString()
-	for key := range m {
-		keys.Insert(key)
+	keys := make([]string, 0, len(m))
+	var grow int
+	for k, v := range m {
+		keys = append(keys, k)
+		// why + 2? (=,\n)
+		// this is just a vague size to avoid frequent requests for space
+		// why is it 3*len(v)/2, from strconv.quoteWith
```
Comment (SergeyKanzhelev) [inline pkg/fieldpath/fieldpath.go:38]: I don't understand the `3*len(v)/2` even after this comment. Can you expand?

File: pkg/fieldpath/fieldpath.go [hunk 4]
```diff
@@ -18,26 +18,38 @@ package fieldpath
...
+			dst = append(dst, '\n')
+		}
+		dst = append(dst, key...)
+		dst = append(dst, '=')
+		dst = strconv.AppendQuote(dst, m[key])
 	}
-	fmtStr = strings.TrimSuffix(fmtStr, "\n")
-
-	return
+	return string(dst)
```
Comment (SergeyKanzhelev) [inline pkg/fieldpath/fieldpath.go:52]: was the original version trimmed the last newline?

File: pkg/fieldpath/fieldpath_test.go [hunk 5]
```diff
@@ -24,6 +24,22 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+func BenchmarkFormatMap(b *testing.B) {
```
Comment (SergeyKanzhelev) [inline pkg/fieldpath/fieldpath_test.go:27]: how this benchmark func will be used?

Comment (aimuz) [reply pkg/fieldpath/fieldpath_test.go:27]:
```
go test -v -bench=BenchmarkFormatMap -benchmem ./pkg/fieldpath -count 5 > new.out
```

```
git checkout master pkg/fieldpath/fieldpath.go
```

```
go test -v -bench=BenchmarkFormatMap -benchmem ./pkg/fieldpath -count 5 > old.out
```

```
benchstat old.out new.out 
name          old time/op    new time/op    delta
FormatMap-12    1.46µs ±17%    0.42µs ±23%  -71.07%  (p=0.008 n=5+5)

name          old alloc/op   new alloc/op   delta
FormatMap-12      576B ± 0%      232B ± 0%  -59.72%  (p=0.008 n=5+5)

name          old allocs/op  new allocs/op  delta
FormatMap-12      15.0 ± 0%       4.0 ± 0%  -73.33%  (p=0.008 n=5+5)
```

File: pkg/fieldpath/fieldpath.go (hunk 3 above)
Comment (liggitt) [reply pkg/fieldpath/fieldpath.go:38]: I'm not convinced we need this estimation. This function is used to append labels, whose values don't need escaping, and annotations, whose values usually don't need escaping.

Allocating a buffer with enough space for len(k) + len(v) + 3 (for `=`, `"` and `"`) and writing into the buffer would be much simpler to implement and read.

**Comment by aimuz** [comment]
Fixed

File: pkg/fieldpath/fieldpath_test.go [hunk 6]
```diff
@@ -268,3 +284,60 @@ func TestSplitMaybeSubscriptedPath(t *testing.T) {
...
+				m: map[string]string{
+					"beta.kubernetes.io/os":                 "linux",
+					"kubernetes.io/arch":                    "amd64",
+					"kubernetes.io/hostname":                "master01",
+					"kubernetes.io/os":                      "linux",
+					"node-role.kubernetes.io/control-plane": "true",
+					"node-role.kubernetes.io/master":        "true",
+				},
+			},
+			wantFmtStr: "beta.kubernetes.io/os=\"linux\"\nkubernetes.io/arch=\"amd64\"\nkubernetes.io/hostname=\"master01\"\nkubernetes.io/os=\"linux\"\nnode-role.kubernetes.io/control-plane=\"true\"\nnode-role.kubernetes.io/master=\"true\"",
```
Comment (liggitt) [inline pkg/fieldpath/fieldpath_test.go:317]: this is really difficult to evaluate for correctness... I'd suggest using ` quoted strings, like this:

```suggestion
			wantFmtStr: `beta.kubernetes.io/os="linux"
kubernetes.io/arch="amd64"
kubernetes.io/hostname="master01"
kubernetes.io/os="linux"
node-role.kubernetes.io/control-plane="true"
node-role.kubernetes.io/master="true"`,
```

same comment applies below

File: pkg/fieldpath/fieldpath_test.go [hunk 7]
```diff
@@ -268,3 +284,60 @@ func TestSplitMaybeSubscriptedPath(t *testing.T) {
...
+		{
+			name: "annotation",
+			args: args{
+				m: map[string]string{
+					"flannel.alpha.coreos.com/backend-data":                  "{\"VNI\":1,\"VtepMAC\":\"ce:f9:c7:a4:de:64\"}",
+					"flannel.alpha.coreos.com/backend-type":                  "vxlan",
+					"flannel.alpha.coreos.com/kube-subnet-manager":           "true",
+					"flannel.alpha.coreos.com/public-ip":                     "192.168.19.160",
+					"management.cattle.io/pod-limits":                        "{\"cpu\":\"11400m\",\"memory\":\"7965Mi\"}",
+					"management.cattle.io/pod-requests":                      "{\"cpu\":\"2482m\",\"memory\":\"7984Mi\",\"pods\":\"26\"}",
```
Comment (liggitt) [inline pkg/fieldpath/fieldpath_test.go:328]: use ` quoted strings to avoid excessive backslash escaping

File: pkg/fieldpath/fieldpath.go [hunk 8]
```diff
@@ -18,26 +18,36 @@ package fieldpath
...
 	// output with keys in sorted order to provide stable output
-	keys := sets.NewString()
-	for key := range m {
-		keys.Insert(key)
+	keys := make([]string, 0, len(m))
+	var grow int
+	for k, v := range m {
+		keys = append(keys, k)
+		// why add 3: (for =, " and ")
+		grow += len(k) + len(v) + 3
```
Comment (liggitt) [inline pkg/fieldpath/fieldpath.go:37]: actually, I guess this should add 4 to account for the newline as well

Comment (aimuz) [reply pkg/fieldpath/fieldpath.go]: Fixed

It does miss the line break

File: pkg/fieldpath/fieldpath_test.go (hunk 7 above)
Comment (aimuz) [reply pkg/fieldpath/fieldpath_test.go]: Fixed

File: pkg/fieldpath/fieldpath_test.go (hunk 6 above)
Comment (aimuz) [reply pkg/fieldpath/fieldpath_test.go]: Fixed

**Pull request merged**